odfdom/src/main/java/org/odftoolkit/odfdom/pkg/OdfPackage.java

```diff
@@ -2025,32 +2030,44 @@ private byte[] decryptData(
       String checksum = encryptionDataElement.getChecksumAttribute();
       byte[] salt = Base64Binary.valueOf(saltStr).getBytes();
       byte[] iv = Base64Binary.valueOf(ivStr).getBytes();
-      byte[] passBytes = mOldPwd.getBytes("UTF-8");
-      MessageDigest md = MessageDigest.getInstance("SHA-1");
-      passBytes = md.digest(passBytes);
-      /*
-       * char passChars[] = new char[passBytes.length]; for(int i = 0;
-       * i<passBytes.length; i++){ passChars[i] =
-       * (char)(passBytes[i]|0xFF); } KeySpec spec = new
-       * PBEKeySpec(passChars, salt, 1024, 128); SecretKeyFactory factory
-       * = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1"); SecretKey
-       * skey = factory.generateSecret(spec); byte[] raw =
-       * skey.getEncoded(); SecretKeySpec skeySpec = new
-       * SecretKeySpec(raw, "Blowfish");
-       */
-      byte[] dk = derivePBKDF2Key(passBytes, salt, 1024, 16);
-      SecretKeySpec key = new SecretKeySpec(dk, "Blowfish");
+      byte[] passBytes = mOldPwd.getBytes(StandardCharsets.UTF_8);
+
+      String algorithm = algorithmElement.getAlgorithmNameAttribute();
+      boolean isAES = algorithm.contains("aes256-cbc") || algorithm.contains("AES-256-CBC");
+
+      SecretKeySpec key;
+      Cipher cipher;
+      MessageDigest md;
+
+      if (isAES) {
+        // AES-256-CBC
+        md = MessageDigest.getInstance("SHA-256");
+        passBytes = md.digest(passBytes);
+        PBEParametersGenerator generator = new PKCS5S2ParametersGenerator(new SHA1Digest());
+        generator.init(passBytes, salt, 100000);
+        KeyParameter keyParam = (KeyParameter) generator.generateDerivedParameters(256);
+        key = new SecretKeySpec(keyParam.getKey(), "AES");
+        cipher = Cipher.getInstance("AES/CBC/NoPadding");
```

Comment:
i believe AES uses W3C padding? (whereas Blowfish didn't need padding)

hmm ... i find this page: https://docs.oracle.com/en/java/javase/11/docs/specs/security/standard-names.html

this line probably?

ISO10126Padding 	This padding for block ciphers is described in [5.2 Block Encryption Algorithms](https://www.w3.org/TR/xmlenc-core/#sec-Alg-Block) in the W3C “XML Encryption Syntax and Processing” document.

Reply:
It only worked for me without padding.
I got my info from this StackOverflow posting: [https://stackoverflow.com/questions/62001716/encrypt-a-file-inside-an-ods-archive/62005258?r=Saves_UserSavesList#62005258](https://stackoverflow.com/questions/62001716/encrypt-a-file-inside-an-ods-archive/62005258?r=Saves_UserSavesList#62005258)

Reply:
hmm but that also says to use ISO10126Padding?